Create common release resources for custom tasks

lbernick opened the pull request:
# Changes
This commit moves the release pipeline and task that are copy-pasted in many places in experimental/
to a common location. Individual projects in experimental should be migrated to use this Pipeline and Task
with appropriate parameter values, instead of continuing to copy paste these definitions in multiple places.
This will be used first for the pipeline-in-pod project (the ColocatedPipelineRun custom task).

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- n/a Commit messages includes a project tag in the subject - e.g. [OCI], [hub], [results], [task-loops]

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

Part 1 of #805

lbernick commented:
/assign afrittoli

This will probably need https://github.com/tektoncd/plumbing/pull/1073-- I'm guessing the best way to do this would be to add the new /tekton folder first, then merge the changes to plumbing, then remove each project's individual /tekton folder?

afrittoli commented:
Thanks for starting this @lbernick - that's something I've had in the back of my mind for some time now, but didn't have time to look for a proper solution. Some of the projects in the experimental are custom task controllers, some are generic controllers and some are something else entirely, which means their publish task and release pipeline may look different.

For custom task specifically, have a template for releasing them would be useful also as part of https://github.com/tektoncd/community/blob/main/teps/0071-custom-task-sdk.md

A few issues on this PR:
- some projects have a controller and a webhook, some only a controller image, and this is not reflected in this change, so for instance task-loop will be broken by this
- some projects have special logic in case of missing "kodata" folders - we should include that in the generic task

More comments:
-  a generic tekton folder might be confusing as it does not apply to all projects, perhaps we could at least add some comments to the README to describe which projects are using this and what are the requirements for a projects to be able to use these release resources
- we could have a small tekton/README in each project to mention that they use the shared resources for release
- one small downside of with this approach is that owners of the project don't own anymore the release resources. This is already true however for shared resources in plumbing so perhaps it's ok
- if the publish task can be made generic enough, we could eventually get it in the catalog

lbernick commented:
> Thanks for starting this @lbernick - that's something I've had in the back of my mind for some time now, but didn't have time to look for a proper solution. Some of the projects in the experimental are custom task controllers, some are generic controllers and some are something else entirely, which means their publish task and release pipeline may look different.
> 

Got it-- I've renamed the files in /tekton to indicate that this release Task and Pipeline are specifically for releasing a custom task.

> * some projects have a controller and a webhook, some only a controller image, and this is not reflected in this change, so for instance task-loop will be broken by this

I updated the koparse step in the Task to loop over cmd/* instead of ["webhook" "controller"] if that's what you're referring to

> * some projects have special logic in case of missing "kodata" folders - we should include that in the generic task

This should be in there now

> * a generic tekton folder might be confusing as it does not apply to all projects, perhaps we could at least add some comments to the README to describe which projects are using this and what are the requirements for a projects to be able to use these release resources

Updated the readme to indicate that this is just for custom Tasks

> * we could have a small tekton/README in each project to mention that they use the shared resources for release

Updated each project's readme to link to the TODO of swapping them to the new release pipeline. This way we avoid breaking existing projects, and just test it out with ColocatedPipelineRun.

> * one small downside of with this approach is that owners of the project don't own anymore the release resources. This is already true however for shared resources in plumbing so perhaps it's ok

I think this is probably OK! I like the idea of a common Pipeline and Task for releasing custom tasks and each project can have its own TaskRun and PipelineRun or something.

> * if the publish task can be made generic enough, we could eventually get it in the catalog

SGTM! :)

lbernick commented:
@afrittoli @dibyom @jerop I'd like to use this for release automation for the ColocatedPipelineRun custom task, so I'd like to get it moving forward, mind taking a look?

afrittoli reviewed (approved):
Thank you Lee!

dibyom commented on tekton/README.md:
```diff
@@ -0,0 +1,36 @@
+
+### Service account and secrets
+
+In order to release, these Pipelines expects a service account JSON file to be 
```
Hmmm...at some point we should move these to workload identity. Maybe we should have an issue to track that?

dibyom commented on tekton/publish-customtask.yaml:
```diff
@@ -0,0 +1,187 @@
+      # included in the container image when they're used as a dependency.
+      # Rather than trying to determine which deps have this requirement (and
+      # probably get it wrong), we'll just targz up the whole vendor tree and
+      # include it. As of 9/20/2019, this amounts to about 11MB of additional
```
The last sentence is almost certainly inaccurate - it was true for the pipelines repo at some point. I think we should just remove it

dibyom reviewed (approved):
One one minor thing involving removing a comment which I think can happen in a followup. So,

/lgtm
